### PR TITLE
Refactor toTag/fromTag to follow current conventions.

### PR DESCRIPTION
**Pull request by LambdAurora**
This refactor is done following [the current conventions about serialization](https://github.com/FabricMC/yarn/blob/21w05b/CONVENTIONS.md#serialization).

The refactor has been assisted by a script: https://gist.github.com/LambdAurora/04608dac389e91cb5c41553641f6b8d2#file-tag_nbt_refactor-mjs

Everything has been checked afterwards and should be good.

Partially fixing #1989.

### CODE REVIEW

**Review by Prospector — APPROVED** [review]
I have not yet double checked to make sure you didn't miss any, but I approve what is here so far.

**Review by ghost — APPROVED** [review]
Content of PR itself looks good, I also haven't checked for any missed cases. 

Edit:
Should to/fromTag which take/return a PacketByteBuf / ByteBuf be included in this refactor.
This also seems to have missed to/fromTag which have multiple parameters such as:
`Inventories::toTag` aka `class_1262::method_5426`
`METHOD method_5426 toTag (Lnet/minecraft/class_2487;Lnet/minecraft/class_2371;)Lnet/minecraft/class_2487;`
And also some non-standard methods like:
`CompasItem::writeToTag` aka `class_1759::method_27315`
`METHOD method_27315 writeToTag (Lnet/minecraft/class_5321;Lnet/minecraft/class_2338;Lnet/minecraft/class_2487;)V`

Only two examples, used regex to find ( not perfect `[tT]ag.+\(.+\)`), not sure how many are actually applicable.

**Comment by liach** [comment]
Would be better if you include existing `(from/to/read/write)Nbt` methods and see if they are of the right type.

**Review by jaskarth — APPROVED** [review]

File: mappings/net/minecraft/entity/Entity.mapping [hunk 1]
```diff
@@ -351,7 +351,7 @@ CLASS net/minecraft/class_1297 net/minecraft/entity/Entity
 		ARG 5 z
 	METHOD method_5650 remove (Lnet/minecraft/class_1297$class_5529;)V
 		ARG 1 reason
-	METHOD method_5651 fromTag (Lnet/minecraft/class_2487;)V
+	METHOD method_5651 readNbt (Lnet/minecraft/class_2487;)V
 		ARG 1 tag
 	METHOD method_5652 writeCustomDataToTag (Lnet/minecraft/class_2487;)V
```
Comment (Juuxel) [inline mappings/net/minecraft/entity/Entity.mapping]: These custom data methods also need to use `Nbt`.

**Review by YanisBft — APPROVED** [review]

**Comment by LambdAurora** [comment]
> Would be better if you include existing `(from/to/read/write)Nbt` methods and see if they are of the right type.

@liach I added logging for potential misnames, the script only complained about `net/minecraft/text/NbtText#toNbt` being misnamed but that's just because it can't read generics and the method returns a stream of `CompoundTag`s.

So it should be alright.

File: mappings/net/minecraft/entity/Entity.mapping [hunk 2]
```diff
@@ -366,7 +366,7 @@ CLASS net/minecraft/class_1297 net/minecraft/entity/Entity
 	METHOD method_5660 setSneaking (Z)V
 		ARG 1 sneaking
 	METHOD method_5661 getArmorItems ()Ljava/lang/Iterable;
-	METHOD method_5662 saveToTag (Lnet/minecraft/class_2487;)Z
+	METHOD method_5662 saveToNbt (Lnet/minecraft/class_2487;)Z
```
Comment (Juuxel) [inline mappings/net/minecraft/entity/Entity.mapping]: I think this and `saveSelfToNbt`/`method_5786` need a closer look sometime later. These names are a bit weird, and a later PR really needs to rename them to be more descriptive.

- `saveToNbt` does `this.hasVehicle() ? false : this.saveSelfToNbt(tag)`
- `saveSelfToNbt` checks if there is a removal reason that should save, then puts the entity ID in a compound tag and calls `writeNbt` afterward.

File: mappings/net/minecraft/scoreboard/ScoreboardState.mapping [hunk 3]
```diff
@@ -15,5 +15,5 @@ CLASS net/minecraft/class_273 net/minecraft/scoreboard/ScoreboardState
 		ARG 1 tag
 	METHOD method_1222 displaySlotsToTag (Lnet/minecraft/class_2487;)V
 		ARG 1 tag
-	METHOD method_32481 fromTag (Lnet/minecraft/class_2487;)Lnet/minecraft/class_273;
+	METHOD method_32481 fromNbt (Lnet/minecraft/class_2487;)Lnet/minecraft/class_273;
```
Comment (Juuxel) [inline mappings/net/minecraft/scoreboard/ScoreboardState.mapping]: Those `*From/Tag` methods also need to be renamed to use `read*FromNbt`. (and `displaySlotsToTag` to `writeDisplaySlotsToNbt`)

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: mappings/net/minecraft/scoreboard/ScoreboardState.mapping (hunk 3 above)
Comment (LambdAurora) [reply mappings/net/minecraft/scoreboard/ScoreboardState.mapping]: Should be good now.

**Comment by LambdAurora** [comment]
> Should to/fromTag which take/return a PacketByteBuf / ByteBuf be included in this refactor.
> This also seems to have missed to/fromTag which have multiple parameters such as:
> `Inventories::toTag` aka `class_1262::method_5426`
> `METHOD method_5426 toTag (Lnet/minecraft/class_2487;Lnet/minecraft/class_2371;)Lnet/minecraft/class_2487;`
> And also some non-standard methods like:
> `CompasItem::writeToTag` aka `class_1759::method_27315`
> `METHOD method_27315 writeToTag (Lnet/minecraft/class_5321;Lnet/minecraft/class_2338;Lnet/minecraft/class_2487;)V`
> 
> Only two examples, used regex to find ( not perfect `[tT]ag.+\(.+\)`), not sure how many are actually applicable.

@NinjaPhenix for the packet stuff, I don't know, may be a good idea.
I tried to apply your suggestions, should cover more methods now.

There's definitely some methods missing as the script caught some that couldn't get renamed automatically.

**Review by liach — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: mappings/net/minecraft/structure/StructurePiece.mapping [hunk 4]
```diff
@@ -172,7 +172,7 @@ CLASS net/minecraft/class_3443 net/minecraft/structure/StructurePiece
 		ARG 6 y
 		ARG 7 z
 		ARG 8 state
-	METHOD method_14946 getTag ()Lnet/minecraft/class_2487;
+	METHOD method_14946 getNbt ()Lnet/minecraft/class_2487;
```
Comment (Juuxel) [inline mappings/net/minecraft/structure/StructurePiece.mapping]:
```suggestion
	METHOD method_14946 toNbt ()Lnet/minecraft/class_2487;
```
(this does `writeNbt` with a new compound tag and adds some extra properties to it, including the ID and bounding box)

**Review by Juuxel — CHANGES_REQUESTED** [review]
This should be my final review, nice work on this :tiny_potato:

**Review by Juuxel — APPROVED** [review]

**Comment by YanisBft** [comment]
Is this ready to merge or do you have remaining methods to check @LambdAurora?

**Comment by LambdAurora** [comment]
@YanisBft should be ready to merge

All methods that were renamed have been checked. `ItemStack` might need attention in another PR as I'm not sure how to approach its methods for now.

From #1989 there's also talk about refactoring the whole NBT package but that's not the aim of this PR anyway.

**Comment by modmuss50** [comment]
I think this needs extra reviews as its an impactful refactor.

**Review by i509VCB — COMMENTED** [review]
I recall there was a getTag etc method on item stack, are those in scope for rename or will those stay the same.

**Comment by LambdAurora** [comment]
> I recall there was a getTag etc method on item stack, are those in scope for rename or will those stay the same.

Those stay the same for now as I'm not sure how to rename them for now, especially the `getSubTag` and `getOrCreateSubTag`

**Comment by haykam821** [comment]
I think `getTag` and similar are fine to keep, at least for now. If it was renamed I would use something along the lines of 'data tag' or 'data NBT' instead.

**Comment by liach** [comment]
> 
> 
> I recall there was a getTag etc method on item stack, are those in scope for rename or will those stay the same.

In the nbt format the key for that subtag is `tag` ([minecraftwiki](https://minecraft.gamepedia.com/Player.dat_format/ItemNoSlot)), so it's totally fine. Moreover the scope of this pr is serialization content and that method isn't.

**Comment by i509VCB** [comment]
Hmm okay so serialization is the matter for this pr.

**Review by i509VCB — APPROVED** [review]

**Comment by liach** [comment]
Imo I will just one day pr to add doc to itemstack that it gets the subtag stored in `tag` in the nbt format

**Comment by Prospector** [comment]
I think even `getTag` methods are confusing since that could easily be confused with a method that gets the datapack tags. For things like ItemStack.getSubTag I'd go for ItemStack.getNbtSubTag

**Comment by liach** [comment]
> 
> 
> I think even `getTag` methods are confusing since that could easily be confused with a method that gets the datapack tags. For things like ItemStack.getSubTag I'd go for ItemStack.getNbtSubTag

Out of scope.

In addition, for the tags that are collections of items, etc. They are extrinsic properties than intrinsic properties. `getTag` refers to intrinsic properties and should be really easy to distinguish. Treating these datapack tags as intrinsic properties should always be avoided for best practice (i.e. expect a block/item to be no longer in a tag, expect a tag to be empty, etc.)

**Review by modmuss50 — APPROVED** [review]

**Comment by YanisBft** [comment]
@LambdAurora mind fixing the conflict please?

**Comment by LambdAurora** [comment]
> @LambdAurora mind fixing the conflict please?

@YanisBft done

and note to self: GitHub resolve conflict editor changes pasted tabs to spaces >:(

**Comment by YanisBft** [comment]
Thanks! Is it ready to merge now, even with only 1 additional approval?

**Comment by liach** [comment]
maybe just assign modmuss and let him decide.

**Comment by modmuss50** [comment]
fuck merged into the wrong branch

**Comment by modmuss50** [comment]
no I didnt, im blind 👁️

**Pull request merged**